### PR TITLE
Using filtering parameters and namedQuery in the count query

### LINKED ISSUE #41500: The total count for page links do not considered the complete query
### Describe the bug

I'm using the quarkus-hibernate-orm-rest-data-panache extension to generate CRUD's, and in the endpoint that lists the records, the pagination links are not correct. This only becomes evident when I use a filter and/or custom query. Reviewing the source code of the rest-data-panache extension I discovered the reason:

```java
ResultHandle pageCount = tryBlock.invokeVirtualMethod(
                    ofMethod(resourceMetadata.getResourceClass(), Constants.PAGE_COUNT_METHOD_PREFIX + RESOURCE_METHOD_NAME,
                            int.class, Page.class),
                    resource, page);
```
Class: io.quarkus.rest.data.panache.deployment.methods.ListMethodImplementor.java
Line: 232
Version: 3.11.2

The reason is that this SQL query to obtain the total number of records must also consider the filters and/or custom queries that I apply as part of my query-params, and in this way both queries (for the total and list) will be affected by the filters and/or queries:

```java
return creator.invokeVirtualMethod(
        ofMethod(resourceMetadata.getResourceClass(), "list", isNotReactivePanache() ? List.class : Uni.class,
                Page.class, Sort.class, String.class, Map.class),
        resource, page == null ? creator.loadNull() : page, sort, query, dataParams);
```
Class: io.quarkus.rest.data.panache.deployment.methods.ListMethodImplementor.java
Line: 361
Version: 3.11.2

In this second source code fragment you can see that the query and dataParams parameters are being injected, both parameters should also be used to obtain the total.

#### Records in my database:

 id | name | createdBy
--- | --- | --- |
 1  | Jimy | Nikola
 2  | Jaime | Steve
 3  | Julio | Nikola

#### Scenery 1:
Endpoint consumption without filtering parameters:
https://api.jimyx,dev/users?size=2

Headers obtained:
Link: <https://api.jimyx,dev/users?page=0&size=2>; rel="first"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="last"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="next"

#### Scenery 2:
Endpoint consumption with filtering parameters:
https://api.jimyx,dev/users?size=2&createdBy=Nikola

Headers obtained:
Link: <https://api.jimyx,dev/users?page=0&size=2>; rel="first"
Link: <https://api.jimyx,dev/users?page=0&size=2>; rel="last"

#### Records in my database:

 id | name | createdBy
--- | --- | --- |
 1  | Jimy | Nikola
 2  | Jaime | Steve
 3  | Julio | Nikola

#### Scenery 1:
Endpoint consumption without filtering parameters:
https://api.jimyx,dev/users?size=2

Headers obtained:
Link: <https://api.jimyx,dev/users?page=0&size=2>; rel="first"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="last"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="next"

#### Scenery 2:
Endpoint consumption with filtering parameters:
https://api.jimyx,dev/users?size=2&createdBy=Nikola

Headers obtained:
Link: <https://api.jimyx,dev/users?page=0&size=2>; rel="first"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="last"
Link: <https://api.jimyx,dev/users?page=1&size=2>; rel="next"

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by j1myx**
- Fixes: https://github.com/quarkusio/quarkus/issues/41500

### CODE REVIEW

**Comment by j1myx** [comment]
Hi @geoand , @gsmet . Issue https://github.com/quarkusio/quarkus/issues/41500 fixed!

**Comment by geoand** [comment]
Thanks a lot for the conttibution!

It would be great if we could have some tests for this.

**Comment by geoand** [comment]
Also keep in mind that when everything is done, this will need to be rebased onto main

File: docs/src/main/asciidoc/rest-data-panache.adoc [hunk 1]
```diff
@@ -431,6 +431,8 @@ In this example, we have added a named query to list all the persons that contai
 
 Next, we can set a query param `namedQuery` when listing the entities using the generated resource with the name of the named query that we want to use, for example, calling `http://localhost:8080/people?namedQuery=Person.containsInName&name=ter` would return all the persons which name contains the text "ter".
 
+IMPORTANT: If you have pagination enabled, using JOIN FETCH or others might prevent the query (`namedQuery`) from being reused for the count query. To resolve this you can add an additional query by adding a comma in front of it like this `http://localhost:8080/people?namedQuery=Person.containsInName,Person.countContainsInName&name=ter` which will be used for the pagination count.
```
Comment (Sgitario) [inline docs/src/main/asciidoc/rest-data-panache.adoc:434]: I don't see any tests that exercise/cover this. Can you add one?

Comment (Sgitario) [inline docs/src/main/asciidoc/rest-data-panache.adoc:434]: I don't see any tests that exercise/cover this. Can you add one?
Also, I don't fully understand the limitation you describe here, but perhaps I could understand it better with an example in the test. 
In any case, I don't feel that using an extra comma is super intuitive, so perhaps a new namedQueryForCount query param would be a better solution.

**Review by Sgitario — CHANGES_REQUESTED** [review]
It's being a while since I don't see this code, but the changes look promising and even you improve the readibility of some tests. So, no objections about it, thanks a lot!

Yet I would like to see a test that reproduces the issue that is being addressed here (so the test should be failing without these changes). 
Also, I don't very like the approach of using an extra comma to specify the namedQuery only for the count (see my comment from the review), but I also don't fully follow the limitation that you mention. So, another test that exercises this is necessary from my point of view.

**Comment by j1myx** [comment]
> Hace tiempo que no veo este código, pero los cambios parecen prometedores e incluso mejoras la legibilidad de algunas pruebas. Así que no tengo objeciones al respecto, ¡muchas gracias!
> 
> Sin embargo, me gustaría ver una prueba que reproduzca el problema que se aborda aquí (por lo que la prueba debería fallar sin estos cambios). Además, no me gusta mucho el enfoque de usar una coma adicional para especificar la consulta con nombre solo para el recuento (consulte mi comentario de la revisión), pero tampoco sigo completamente la limitación que menciona. Por lo tanto, otra prueba que ejercite esto es necesaria desde mi punto de vista.

Hi @Sgitario, I got your recommendations. I'm going to implement it right now.

**Pull request closed without merging**